server/renderer.js

```diff
@@ -1,363 +1,172 @@
+					this.templates[file] = this.hbs.compile(contents, { preventIndent: true });
+					this.hbs.registerPartial(file, contents);
+				}))
+			.catch(this.crash);
+	}
+
+	// Clears the cache of rendered pages
+	clearCache() {
+		return fs.mkdir(this.RENDER_DIR) // in case the dir doesn't exist, create it to prevent failure
+			.catch(() => undefined) // if it existed, catch the error
+			.then(() => fs.readdir(this.RENDER_DIR))
+			.then((files) => Promise.all(files.map((file) => fs.unlink(pathjoin(this.RENDER_DIR, file)))));
+	}
+
+	// Prerenders all pages specified in routes
+	prerender() {
+		return Promise.all(routes.filter((route) => route.prerender !== undefined)
+				.map((route) => reduceToPromise(route.prerender, this.db)))
+			.then((paths) => paths.reduce((a, b) => a.concat(b), []))
+			.then((paths) => paths.forEach((path) => this.renderPath(path))); // todo: fix scoping?
+	}
+
+	// Renders a single path, pulling from the cache if available, and storing in the cache if allowed
+	renderPath(path, res, user) {
+		logger.info(`Rendering ${path}`);
+		let cachePath = this.getCachePath(path);
+
+		fs.readFile(cachePath)
+			.then((content) => { // the page was cached
+				logger.ok(`Pulled ${path} from cache`);
+				if (res !== undefined) {
+					res.type("text/html"); // TODO
+					res.send(content);
 				}
-				Promise.all(promises).then(resolve, reject)
 			})
-		})
-	},
-	renderAll: function() {
-		var obj = compileRoutes(this.db)
-		this.routes = obj.routes
-		this.prerender = obj.prerender
-		var promises = []
-		for (var p in this.prerender) {
-			if (!this.prerender[p].options || !this.prerender[p].options.groups) {
-				promises.push(this.renderPage(this.prerender[p].path))
-			} else {
-				var groups = this.prerender[p].options.groups
-				for (var i in groups) {
-					if (groups[i].range) {
-						for (var l = groups[i].range.start; l <= groups[i].range.end; l += 1) {
-							var r = new RegExp("\\{"+i+"\\}","g")
-							var url = this.prerender[p].path.replace(r, l)
-							promises.push(this.renderPage(url))
-						}
-					}
-					if (groups[i].each) {
-						for (var l = 0; l < groups[i].each.length; l++) {
-							var r = new RegExp("\\{"+i+"\\}","g")
-							var url = this.prerender[p].path.replace(r, groups[i].each[l])
-							promises.push(this.renderPage(url))
-						}
-					}
-				}
-			}
-		}
-		return Promise.all(promises)
-	},
-	renderPage: function(url) {
-		logger.info("[render]", "Rendering", url)
-		var loc = path.join(this.__dirname, config.paths.render)
-		for (var i in this.routes) {
-			var m = url.match(i)
-			if (m) {
-				var context = extend(true, {}, this.routes[i])
-				if (context.cache === false) {
-					return new Promise(function (resolve, reject) {resolve()})	
-				}
-				for (var ind = 1; ind < m.length; ind++) {
-					if (context.groups && (ind-1) < context.groups.length) {
-						context[context.groups[ind-1]] = m[ind]
-					}
+			.catch(() => { // it's not cached, so we'll render it
```

Comment:
I would specifiy the error here, because otherwise and unexpected error would at best cause performance slow-downs, and at worst mask a more significant problem.